### PR TITLE
[WIP] Trade execution oneshot

### PR DESCRIPTION
**Pull request by ysv**

### CODE REVIEW

File: bin/trade_execution_oneshot [hunk 1]
```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env ruby
+
+# encoding: UTF-8
+# frozen_string_literal: true
+
+require File.join(ENV.fetch('RAILS_ROOT'), 'config', 'environment')
+
+def daemon(amqp_daemon)
```
Comment (yivo) [inline bin/trade_execution_oneshot:8]: This is not correct way to run daemon. You have to use separate threads. Don't forget to terminate them at the end of main program.

File: bin/trade_execution_oneshot [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+# def current_user
+#   Member.first
+# end
+
+def build_order(attrs)
+  (attrs[:side] == 'sell' ? OrderAsk : OrderBid).new \
+        state:         ::Order::WAIT,
+        member:        @current_user,
+        ask:           @current_market&.base_unit,
```
Comment (yivo) [inline bin/trade_execution_oneshot:25]: You can use ask_unit/bid_unit here.

File: bin/trade_execution_oneshot [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+#   Market.first
+# end
+
+# def current_user
+#   Member.first
+# end
+
+def build_order(attrs)
+  (attrs[:side] == 'sell' ? OrderAsk : OrderBid).new \
+        state:         ::Order::WAIT,
```
Comment (yivo) [inline bin/trade_execution_oneshot:23]: `::` is not needed.

File: bin/trade_execution_oneshot [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    amount: 1,
+    txid: "test1",
+    address: '123',
+    txout: '333',
+    aasm_state: "accepted"
+  }
+end
+
+def create_deposit(attrs)
+  currency = Currency.find_by_id(attrs[:currency])
```
Comment (yivo) [inline bin/trade_execution_oneshot:59]: Never use Currency#find_by_id

File: bin/trade_execution_oneshot [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  "Deposits/#{currency.type}".camelize.constantize.create!(
+    member: @current_user,
+    currency: currency,
+    amount: 2,
+    txid: SecureRandom.hex(4),
+    address: SecureRandom.hex(4),
+    txout: SecureRandom.hex(4))
```
Comment (yivo) [inline bin/trade_execution_oneshot:66]: This is not valid value

File: bin/trade_execution_oneshot [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+end
+
+def create_deposit(attrs)
+  currency = Currency.find_by_id(attrs[:currency])
+  "Deposits/#{currency.type}".camelize.constantize.create!(
+    member: @current_user,
+    currency: currency,
+    amount: 2,
+    txid: SecureRandom.hex(4),
+    address: SecureRandom.hex(4),
```
Comment (yivo) [inline bin/trade_execution_oneshot:65]: Use Faker

File: bin/trade_execution_oneshot [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+  }
+end
+
+def create_deposit(attrs)
+  currency = Currency.find_by_id(attrs[:currency])
+  "Deposits/#{currency.type}".camelize.constantize.create!(
+    member: @current_user,
+    currency: currency,
+    amount: 2,
+    txid: SecureRandom.hex(4),
```
Comment (yivo) [inline bin/trade_execution_oneshot:64]: Use Faker

File: bin/trade_execution_oneshot [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    txid: "test1",
+    address: '123',
+    txout: '333',
+    aasm_state: "accepted"
+  }
+end
+
+def create_deposit(attrs)
+  currency = Currency.find_by_id(attrs[:currency])
+  "Deposits/#{currency.type}".camelize.constantize.create!(
```
Comment (yivo) [inline bin/trade_execution_oneshot:60]: Use `deposits/#{}` ...

File: bin/trade_execution_oneshot [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    member: @current_user,
+    currency: currency,
+    amount: 2,
+    txid: SecureRandom.hex(4),
+    address: SecureRandom.hex(4),
+    txout: SecureRandom.hex(4))
+    .tap { |d| d.accept! }
```
Comment (yivo) [inline bin/trade_execution_oneshot:67]: Use tap(&:accept!)

File: bin/trade_execution_oneshot [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    address: '123',
+    txout: '333',
+    aasm_state: "accepted"
+  }
+end
+
+def create_deposit(attrs)
+  currency = Currency.find_by_id(attrs[:currency])
+  "Deposits/#{currency.type}".camelize.constantize.create!(
+    member: @current_user,
```
Comment (yivo) [inline bin/trade_execution_oneshot:61]: Align attributes please

File: bin/trade_execution_oneshot [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    ord_type: :limit,
+    price:     1,
+    volume:    1
+  }
+end
+
+def deposit_attrs
+  {
+    member:   current_user,
+    currency: Currency.find_by_id(:btc),
```
Comment (yivo) [inline bin/trade_execution_oneshot:49]: Never combine find_by_id and currency in this context

File: bin/trade_execution_oneshot [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    amount: 1,
+    txid: "test1",
```
Comment (yivo) [inline bin/trade_execution_oneshot:51]: Don't mix quotes

File: bin/trade_execution_oneshot [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+def order_attrs
+  {
+    ord_type: :limit,
+    price:     1,
+    volume:    1
+  }
+end
+
+def deposit_attrs
```
Comment (yivo) [inline bin/trade_execution_oneshot:46]: What this is needed for?

File: bin/trade_execution_oneshot [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+# frozen_string_literal: true
+
+require File.join(ENV.fetch('RAILS_ROOT'), 'config', 'environment')
+
+def daemon(amqp_daemon)
+  command = "bundle exec ruby lib/daemons/amqp_daemon.rb #{amqp_daemon}"
+  spawn command
+end
+
+# def current_market
```
Comment (yivo) [inline bin/trade_execution_oneshot:13]: Remove comments

File: bin/trade_execution_oneshot [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+ask_unit = 'btc'
+@current_market = Market.find_by(bid_unit: bid_unit, ask_unit: ask_unit)
+@current_user = Member.last
+deposits = [bid_unit, ask_unit].map { |currency| create_deposit(currency: currency) }
+orders = %w[sell buy].map { |side| create_order(order_attrs.merge(side: side)) }
+#binding.pry
+pids = %w[matching trade_executor].map { |d| daemon(d) }
+
+sleep 5
+# # 10.times do
```
Comment (yivo) [inline bin/trade_execution_oneshot:81]: Remove comments

File: bin/trade_execution_oneshot [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+
+end
+
+bid_unit = 'usd'
+ask_unit = 'btc'
+@current_market = Market.find_by(bid_unit: bid_unit, ask_unit: ask_unit)
+@current_user = Member.last
+deposits = [bid_unit, ask_unit].map { |currency| create_deposit(currency: currency) }
+orders = %w[sell buy].map { |side| create_order(order_attrs.merge(side: side)) }
+#binding.pry
```
Comment (yivo) [inline bin/trade_execution_oneshot:77]: Remove comments

File: bin/trade_execution_oneshot [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+    txid: SecureRandom.hex(4),
+    address: SecureRandom.hex(4),
+    txout: SecureRandom.hex(4))
+    .tap { |d| d.accept! }
+
+end
+
+bid_unit = 'usd'
+ask_unit = 'btc'
+@current_market = Market.find_by(bid_unit: bid_unit, ask_unit: ask_unit)
```
Comment (yivo) [inline bin/trade_execution_oneshot:73]: Use Market#find

**Review by yivo — CHANGES_REQUESTED** [review]

**Comment by mod** [comment]
@yssavchuk  Can we move this to peatio-core

**Pull request closed without merging**